File: javascript/2015/day07b/day07b.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const process = require('process');

function usage() {
    const progname = path.basename(process.argv[1]);
    console.log(`usage: ${progname} <input file>`);
    process.exit(1);
}

function processFile(filename) {
    const operations = new Map();
    var cache = new Map();
    const data = fs.readFileSync(filename, 'utf8');
    const lines = data.trim().split('\n');
    for ( const line of lines ) {
        var match = line.match(/^(\d+|\w+) -> (\w+)$/);
        if ( match != null ) {
            const [_, src, dest] = match;
            operations.set(dest, { operator: 'ASSIGN', src: src });
        }
        match = line.match(/NOT (\d+|\w+) -> (\w+)/);
        if ( match != null ) {
            const [_, src, dest] = match;
            operations.set(dest, { operator: 'NOT', src: src });
        }
        match = line.match(/(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)/);
        if ( match != null ) {
            const [_, src1, op, src2, dest] = match;
            operations.set(dest, { operator: op, src1: src1, src2: src2 });
        }
        match = line.match(/(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)/);
        if ( match != null ) {
            const [_, src, op, amtStr, dest] = match;
            operations.set(dest, { operator: op, src: src, amt: parseInt(amtStr) });
        }
    }
    const a = eval(operations, cache, 'a');
    operations.set('b', { operator: 'ASSIGN', src: a.toString() });
    cache = new Map();
    return eval(operations, cache, 'a');
}

function eval(ops, cache, expr) {
    if ( Number.isInteger(Number(expr)) ) {
        return Number(expr);
    }
    else if ( cache.has(expr) ) {
        return cache.get(expr);
    } else {
        let r = 0;
        let a = 0;
        let b = 0;
        const op = ops.get(expr);
        switch ( op.operator ) {
        case 'ASSIGN':
            a = eval(ops, cache, op.src);
            r = a;
            break;
        case 'NOT':
            a = eval(ops, cache, op.src);
            r = ~a;
            break;
        case 'AND':
            a = eval(ops, cache, op.src1);
            b = eval(ops, cache, op.src2);
            r = a & b;
            break;
        case 'OR':
            a = eval(ops, cache, op.src1);
            b = eval(ops, cache, op.src2);
            r = a | b;
            break;
        case 'LSHIFT':
            a = eval(ops, cache, op.src);
            r = a << op.amt;
            break;
        case 'RSHIFT':
            a = eval(ops, cache, op.src);
            r = a >> op.amt;
            break;
        }
        const masked = r & 0xffff;
        cache.set(expr, masked);
        return masked;
    }
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    const filename = process.argv[2];
    const result = processFile(filename);
    console.log(`result = ${result}`);
}

main()

File: ruby/2015/day07b/day07b.rb
#!/usr/bin/env ruby

class Operation
  attr_reader :op, :operands

  def initialize(op, *operands)
    @op = op       # :assign, :not, :and, :or, :lshift, :rshift
    @operands = operands
  end
end

def usage
    progname = File.basename($0)
    puts "usage: #{progname} <input file>"
    exit 1
end

def process(filename)
    operations = {}
    cache = {}
    File.foreach(filename) do |line|
        case line
        when /^(\d+|\w+) -> (\w+)$/
            src, dest = $1, $2
            operations[dest] = Operation.new(:assign, src)
        when /NOT (\d+|\w+) -> (\w+)/
            src, dest = $1, $2
            operations[dest] = Operation.new(:not, src)
        when /(\d+|\w+) (AND|OR) (\d+|\w+) -> (\w+)/
            src1, op_str, src2, dest = $1, $2, $3, $4
            op = (op_str == "AND" ? :and : :or)
            operations[dest] = Operation.new(op, src1, src2)
        when /(\d+|\w+) (LSHIFT|RSHIFT) (\d+) -> (\w+)/
            src, op_str, amt, dest = $1, $2, $3.to_i, $4
            op = (op_str == "LSHIFT" ? :lshift : :rshift)
            operations[dest] = Operation.new(op, src, amt)
        end
    end
    a = evaluate(operations, cache, 'a')
    operations["b"] = Operation.new(:assign, a.to_s)
    cache = {}
    evaluate(operations, cache, 'a')
end

def evaluate(ops, cache, expr)
  return expr.to_i if expr =~ /^\d+$/
  return cache[expr] if cache.key?(expr)
  op = ops.fetch(expr)
  r = case op.op
      when :assign
        evaluate(ops, cache, op.operands[0])
      when :not
        ~evaluate(ops, cache, op.operands[0])
      when :and
        a = evaluate(ops, cache, op.operands[0])
        b = evaluate(ops, cache, op.operands[1])
        a & b
      when :or
        a = evaluate(ops, cache, op.operands[0])
        b = evaluate(ops, cache, op.operands[1])
        a | b
      when :lshift
        a = evaluate(ops, cache, op.operands[0])
        a << op.operands[1]
      when :rshift
        a = evaluate(ops, cache, op.operands[0])
        a >> op.operands[1]
      else
        0
      end
  cache[expr] = r & 0xffff
end

def main
    usage if ARGV.length < 1
    filename = ARGV[0]
    result = process(filename)
    puts "result = #{result}"
end

main if __FILE__ == $0
